### PR TITLE
Honor RAILS_ENV/BP_BUNDLE_WITHOUT to skip dev/test gems from build cache

### PR DESCRIPTION
**Pull request by vishal24367**
Fixes: https://github.com/neetozone/neeto-deploy-web/issues/7146

## Summary

The `build-gems` cache layer was carrying every gem in the `Gemfile`, including `:development` and `:test` groups, even though the launch image correctly excludes them. Inspection of `neeto-planner-web-staging`'s `:cache` ECR image (3.04 GB uncompressed) showed:

| Bucket | Size (uncompressed) |
|---|---|
| `wkhtmltopdf-binary` (`:development, :test` group, 2 cached versions) | 1,134 MB |
| Other dev/test gems (brakeman, faker, rbs, prism, rubocop, ruby_parser, …) | ~152 MB |
| **Total dev/test pollution** | **~1,286 MB** |

The exporter still has to extract + gzip + SHA256 the entire 2.65 GB compressed `build-gems` layer on every build (~85 s/build) — even when nothing changed.

## Root cause

In `build.go`, the **build-layer** install runs with no `without` config, while the **launch-layer** install hardcodes `\"without\": \"development:test\"`. Once dev/test gems land in the build-gems cache, they stay there forever — subsequent builds \"Reuse cached layer\" without re-installing.

## Change

Adds two env knobs honored by **both** layer installs:

- **`BP_BUNDLE_WITHOUT`** — explicit override (highest priority)
- **`RAILS_ENV` / `RACK_ENV`** — derives a default:
  - `production` | `staging` → `\"development:test\"`
  - `development` → `\"production:test:staging:heroku\"`
  - `test` → `\"production:development:staging:heroku\"`
  - anything else / unset → `\"\"` (preserves existing behaviour)

Defaults preserve back-compat: when neither knob is set, the build layer still installs everything and the launch layer still strips dev/test.

## Files

- `environment.go` — new `BundleWithout` field + parsing logic.
- `build.go` — both layer install sites now honor `environment.BundleWithout`.
- `environment_test.go` — 9 new cases (BP_ explicit, BP_ vs RAILS_ENV precedence, each RAILS_ENV value, RACK_ENV fallback, RAILS_ENV wins over RACK_ENV).
- `build_test.go` — 2 new contexts: build-only `BundleWithout`, and combined build+launch `BundleWithout` (asserts launch invocation overrides hardcoded default).
- `buildpack.toml` — `0.8.18` → `0.9.0`.

## Versioning

`0.9.0` (minor — new env-driven feature, defaults preserve back-compat).

## Test plan

- [x] All 11 new unit tests pass.
- [x] No regression vs `main`: pre-existing failures unchanged (9 failures on both `main` and this branch — Detect tests + launch-only Build tests broken since `0b9a148`, unrelated to this change).
- [x] `go build ./...` succeeds.
- [x] Multi-arch buildpack image published to ECR as `348674388966.dkr.ecr.us-east-1.amazonaws.com/neeto-deploy/paketo/buildpack/bundle-install:0.9.0` (manifest list w/ `0.9.0-amd64` + `0.9.0-arm64`).

## Expected savings

- ~600 MB compressed reduction in the build-gems layer
- ~25–30 s/build saved on EXPORT phase
- Same fix benefits **every** Ruby app on neeto-deploy that imports `Gemfile.common.rb` (the `wkhtmltopdf-binary` placement in `:development, :test` is shared across products)

## Out of scope (separate follow-up)

Wiring the new buildpack into the build pipeline — either rebuild the builder image (`paketo-builders/jammy:v1`, etc.) or override at the slug-compiler level via `DEFAULT_BUILDPACKS` `--buildpack` flag in `neeto-deploy-slug-compiler-web/app/services/slug_compiler/pack_build_service.rb`.

Refs: neetozone/neeto-deploy-web#7146

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: environment.go [hunk 1]
```diff
@@ -8,10 +8,31 @@ import (
 
 type Environment struct {
 	KeepGemExtensionBuildFiles bool
+	// BundleWithout is the colon-separated list of Bundler groups to exclude
+	// from `bundle install` in BOTH the build and launch layers. When empty,
+	// the build layer installs all groups (legacy behaviour) and the launch
+	// layer falls back to its hardcoded "development:test" default.
+	BundleWithout string
```
[the remark this answers is not in the document]
Comment (vishal24367) [reply environment.go:15]: Won't fix — not valid for our use case.

The "explicitly empty" semantics this comment proposes (`BP_BUNDLE_WITHOUT=` overriding the RAILS_ENV-derived `development:test` to install ALL gems including dev/test in the launch image) describes a scenario that doesn't apply to neeto-deploy:

1. Every neeto product runs `RAILS_ENV=production` (or `staging`) at runtime; no production deploy ever wants `:development` or `:test` gems shipped.
2. The pre-existing launch-layer hardcoded default (`without development:test`) already enforced this; we're preserving that default exactly when `BP_BUNDLE_WITHOUT` is unset.
3. Adding `*string` would force every consumer (and every test) to nil-check, adding boilerplate for an edge case we can't justify with any deploy in the org.

If a real use case for "install everything in launch" emerges later, we can revisit. Documented the empty-as-unset semantics in the `environment.go` docstring already.

File: build.go [hunk 2]
```diff
@@ -123,10 +123,14 @@ func Build(
 				logger.Process("Executing build environment install process")
 
 				duration, err := clock.Measure(func() error {
-					return installProcess.Execute(context.WorkingDir, layer.Path, map[string]string{
+					buildConfig := map[string]string{
 						"path":  layer.Path,
 						"clean": "true",
-					}, environment.KeepGemExtensionBuildFiles)
+					}
+					if environment.BundleWithout != "" {
```
[the remark this answers is not in the document]
Comment (vishal24367) [reply build.go:130]: Won't fix — not valid for our use case, and confirmed by codebase audit.

The bot's concern ("apps may rely on dev gems at build time") is theoretical. I checked our actual build pipeline:

- The only buildpack that runs Ruby code during BUILD is [`neeto-deploy-paketo-rails-assets-buildpack`](../../../neeto-deploy-paketo-rails-assets-buildpack), which executes `rails assets:precompile` with `RAILS_ENV=production` set explicitly.
- No buildpack in `neeto-deploy-paketo-build-system/ruby/` runs brakeman, rubocop, factory_bot, faker, or any other dev/test gem during build.
- Inspection of the actual `:cache` ECR image for `neeto-planner-web-staging` confirmed that dev gems were *present but unused* — they were leaked from the initial `bundle install` and never re-installed since.

Excluding dev/test from BUILD when `RAILS_ENV=production` is exactly the desired behaviour for our pipeline — it's the entire ~600 MB / ~25-30 s/build saving this PR is targeting. Applying derived `without` only to launch (the bot's suggested fix) would mean *zero* automatic savings, requiring every app to set `BP_BUNDLE_WITHOUT` explicitly.

Apps in the broader CNB community with unusual build-time dependencies on dev gems can opt out by either (a) not setting `RAILS_ENV` during build, or (b) setting `BP_BUNDLE_WITHOUT` to a value that excludes only what they actually want excluded.
